python/cudf/cudf/core/column_accessor.py

```diff
@@ -253,16 +255,26 @@ def _grouped_data(self) -> abc.MutableMapping:
         else:
             return self._data
 
-    def _clear_cache(self):
+    def _clear_cache(self, old_ncols: int, new_ncols: int):
+        """
+        Clear cached attributes.
+
+        Parameters
+        ----------
+        old_ncols: int
+            len(self._data) before self._data was modified
+        new_ncols: int
+            len(self._data) after self._data was modified
+        """
         cached_properties = ("columns", "names", "_grouped_data")
         for attr in cached_properties:
             try:
                 self.__delattr__(attr)
             except AttributeError:
                 pass
 
-        # nrows should only be cleared if no data is present.
-        if len(self._data) == 0 and hasattr(self, "nrows"):
+        # nrows should only be cleared if empty before/after the op.
+        if (old_ncols == 0 or new_ncols == 0) and hasattr(self, "nrows"):
```

Comment:
Since `nrows` is a `cached_property` `hasattr(self, "nrows")` will always return True (since it evaluates the property if it isn't already extant).

I think you want:
```
if old_ncols == 0 or new_cols == 0:
    try:
        del self.nrows
    except AttributeError:
        pass
```

`self.__delattr__("nrows")` works identically.

Aside: your logic is not quite right (unless there is a precondition that `old_ncols != new_ncols`) since you want `old_ncols == 0 XOR new_ncols == 0` to flag the change from empty to non-empty or vice versa.

Reply:
Nice catch on xor. I _think_ that precondition generally holds since this method is called when `._data` is modified to add/remove labels, but can't hurt to xor anyways